File: server/src/main/java/com/gitforge/vcs/ref/BranchName.java
package com.gitforge.vcs.ref;

import java.util.Set;

/**
 * The single place branch naming rules are defined.
 *
 * <p>A branch name becomes a path under {@code refs/heads}, so an unvalidated
 * name is a filesystem write primitive: {@code ../../objects/ab/cdef} would let
 * a caller overwrite an immutable object through the mutable ref API. Validation
 * here is the first of two defences; {@link FileSystemRefStore} independently
 * confirms that the resolved path stays inside the refs directory, so a rule
 * missed here still cannot escape.
 *
 * <p>Slashes are permitted as hierarchy separators, so {@code feature/login} is a
 * valid name that nests one directory deep.
 */
public final class BranchName {

    /** Characters that carry meaning in revision syntax, or that break paths. */
    private static final Set<Character> FORBIDDEN_CHARACTERS =
            Set.of('~', '^', ':', '?', '*', '[', ']', '\\', '"', '\'', '<', '>', '|', ' ');

    /** Reserved because it names the special HEAD reference. */
    private static final String RESERVED_HEAD = "HEAD";

    private static final String LOCK_SUFFIX = ".lock";

    private BranchName() {
    }

    /**
     * Returns {@code name} unchanged if it is a legal branch name.
     *
     * @throws RefException if the name is unsafe or malformed
     */
    public static String validate(String name) {
        if (name == null || name.isBlank()) {
            throw new RefException("Branch name must not be empty");
        }
        if (name.equals(RESERVED_HEAD)) {
            throw new RefException("Branch name '" + RESERVED_HEAD + "' is reserved");
        }
        if (name.startsWith("/") || name.endsWith("/")) {
            throw new RefException("Branch name must not start or end with '/': " + name);
        }
        if (name.contains("//")) {
            throw new RefException("Branch name must not contain an empty segment: " + name);
        }
        // Catches absolute Windows paths such as C:\work before the character
        // scan reports something less specific.
        if (name.length() > 1 && name.charAt(1) == ':') {
            throw new RefException("Branch name must not be an absolute path: " + name);
        }

        for (char character : name.toCharArray()) {
            if (character < 0x20 || character == 0x7F) {
                throw new RefException("Branch name must not contain control characters: " + name);
            }
            if (FORBIDDEN_CHARACTERS.contains(character)) {
                throw new RefException("Branch name must not contain '" + character + "': " + name);
            }
        }
        if (name.contains("@{")) {
            throw new RefException("Branch name must not contain '@{': " + name);
        }

        for (String segment : name.split("/", -1)) {
            validateSegment(segment, name);
        }
        return name;
    }

    private static void validateSegment(String segment, String fullName) {
        if (segment.isEmpty()) {
            throw new RefException("Branch name must not contain an empty segment: " + fullName);
        }
        // The escape vector: these would resolve outside refs/heads.
        if (segment.equals(".") || segment.equals("..")) {
            throw new RefException("Branch name must not contain '.' or '..' segments: " + fullName);
        }
        if (segment.startsWith(".")) {
            throw new RefException("Branch name segments must not start with '.': " + fullName);
        }
        if (segment.startsWith("-")) {
            throw new RefException("Branch name segments must not start with '-': " + fullName);
        }
        if (segment.endsWith(LOCK_SUFFIX)) {
            throw new RefException("Branch name segments must not end with '" + LOCK_SUFFIX + "': " + fullName);
        }
    }
}

File: frontend/src/utils/branchName.js
/**
 * Mirrors com.gitforge.vcs.ref.BranchName.
 *
 * A branch name becomes a path under refs/heads on the server, which is why the
 * rules are strict there. Repeating them here does not make the client an
 * authority — the server validates independently and its message is shown
 * verbatim if it disagrees — it only means an obviously bad name is reported
 * while typing instead of after a round trip.
 */

/** Characters that carry meaning in revision syntax, or that break paths. */
const FORBIDDEN = new Set(["~", "^", ":", "?", "*", "[", "]", "\\", '"', "'", "<", ">", "|", " "]);

/** Matches the @Size(max = 255) bound on CreateBranchRequest. */
export const MAX_BRANCH_NAME_LENGTH = 255;

export function validateBranchName(rawName) {
  const name = rawName ?? "";

  if (!name.trim()) return "A branch name is required.";
  if (name.length > MAX_BRANCH_NAME_LENGTH) {
    return `Must be ${MAX_BRANCH_NAME_LENGTH} characters or fewer.`;
  }
  if (name === "HEAD") return "'HEAD' is reserved.";
  if (name.startsWith("/") || name.endsWith("/")) return "Cannot start or end with '/'.";
  if (name.includes("//")) return "Cannot contain an empty segment.";
  // Catches an absolute Windows path such as C:\work before the character scan
  // reports something less specific.
  if (name.length > 1 && name[1] === ":") return "Cannot be an absolute path.";
  if (name.includes("@{")) return "Cannot contain '@{'.";

  for (const character of name) {
    const code = character.codePointAt(0);
    if (code < 0x20 || code === 0x7f) return "Cannot contain control characters.";
    if (FORBIDDEN.has(character)) return `Cannot contain '${character}'.`;
  }

  for (const segment of name.split("/")) {
    const problem = validateSegment(segment);
    if (problem) return problem;
  }
  return null;
}

function validateSegment(segment) {
  if (!segment) return "Cannot contain an empty segment.";
  if (segment === "." || segment === "..") return "Cannot contain '.' or '..' segments.";
  if (segment.startsWith(".")) return "Segments cannot start with '.'.";
  if (segment.startsWith("-")) return "Segments cannot start with '-'.";
  if (segment.endsWith(".lock")) return "Segments cannot end with '.lock'.";
  return null;
}
